### ISSUE TITLE
Handle CLUSTERDOWN and TRYAGAIN

### ISSUE THREAD

**Issue by nussjustin** [mediocregopher/radix#136]
When debugging #126 I noticed that there are 2 temporary cluster errors that radix currently does not handle:

* `CLUSTERDOWN`, which is returned when not all slots can be served and
* `TRYAGAIN` which is returned when trying multi key operations on a slot that is getting resharded

Currently radix returns these errors as is, but I propose we instead try to handle them similar to how we handle `ASK` and `MOVED` by just retrying the command (if it's retryable).

Also we should trace both of them via the new `trace` subpackage.

* For `TRYAGAIN` we can simply add a new field to `trace.ClusterRedirected`.
* For `CLUSTERDOWN`, which is a global state, I propose we add a new callback `Down` that receives the current topology as known to the `*Cluster` as a list of `trace.ClusterNodeInfo`.

**Comment by nussjustin** [comment]
@mediocregopher Wanted to get feedback on the idea before opening a PR. I've already implemented the proposed changes, including tests for the traced values.

My changes are [here](https://github.com/mediocregopher/radix/compare/master...nussjustin:cluster-failover-handling?expand=1).

When running the benchmark from #126 with my changes the client correctly tries to retry on `CLUSTERDOWN` and by adding the following to the `NewCluster` call basically all `CLUSTERDOWN` errors are handled without the benchmark code noticing:

```go
	ClusterWithTrace(trace.ClusterTrace{
		Down: func(trace.ClusterDown) {
			log.Println("cluster down!")
			time.Sleep(time.Second)
		},
	}
```

The benchmark was run with 100msec +/- 1000msec latency between all cluster nodes and the client. The latency was simulated via

```
sudo tc qdisc replace dev lo root handle 1:0 netem delay 100msec 1000msec 25% loss 12.5%
```

and can be removed via

```
sudo tc qdisc del dev lo root
```

Based on this I think there could be value for adding a way to let the `*Cluster` wait before retrying on `CLUSTERDOWN` but I haven't tried implementing this and wanted to keep the proposed changes small. The above hack using the new `ClusterTrace.Done` callback and `time.Sleep(time.Second)` works, but is just a hack and having a built in solution would be more helpfull.

**Comment by mediocregopher** [comment]
So the spec says this about TRYAGAIN:

> Operations on keys that don't exist or are - during the resharding - split between the source and destination nodes, will generate a -TRYAGAIN error. The client can try the operation after some time, or report back the error.

I'm guessing my original reasoning for not handling TRYAGAIN (if I knew about it at all) was that it's not clear how long a reshard will take, and so retrying again in a loop without any kind of wait in between wouldn't necessarily do anything except slow things down. It might be better to just keep returning it. And if it's being returned directly to the user when it happens, is it necessary to trace it?

Some special handling for CLUSTERDOWN is an interesting idea. The changes you have call `traceClusterDown` inline with the user's call to `Do`, so the `time.Sleep` is just pausing the user's code. Do you think it would be better to have Cluster delay _all_ calls to `Do` as soon as a CLUSTERDOWN has been seen? Theoretically that would prevent the most disruption to the user, but it's also fairly drastic.

I'm not sure if we should trace on every occurrence of CLUSTERDOWN, since again if it's getting returned directly to the user the trace isn't giving them any information they don't already have. But if there's something Cluster is doing in the background when a CLUSTERDOWN occurs then I think _that_ would be worth tracing.

Thanks so much for all your help debugging these issues (there's so many!), I've been busy with a new job so it's definitely nice to have help :)

**Comment by nussjustin** [comment]
> I'm guessing my original reasoning for not handling TRYAGAIN (if I knew about it at all) was that it's not clear how long a reshard will take, and so retrying again in a loop without any kind of wait in between wouldn't necessarily do anything except slow things down. It might be better to just keep returning it. And if it's being returned directly to the user when it happens, is it necessary to trace it?

Maybe we could optionally allow users to set a wait time after which we retry if possible? By default this could be disabled, so we would get the current behavior.

> Do you think it would be better to have Cluster delay all calls to Do as soon as a CLUSTERDOWN has been seen? Theoretically that would prevent the most disruption to the user, but it's also fairly drastic.

Yeah, I also thought about that. We could use a short default wait and allow users to change or disable the waiting. I'd guess that most users would prefer a short wait followed by a retry to just returning the error by default.

Basically on `CLUSTERDOWN` we would track the state as down, and retry once. In `doInner` we could than just check the state at the beginning and sleep for the configured time (if enabled).

> I'm not sure if we should trace on every occurrence of CLUSTERDOWN, since again if it's getting returned directly to the user the trace isn't giving them any information they don't already have. 

It allows for easier tracking, since it can be easier handled in one place. The errors returned by `Do` would need to be manually inspected for `CLUSTERDOWN` in some way.

We could also trace it only once if we track the cluster state (and than maybe just trace whether the cluster is down or not).

**Comment by nussjustin** [comment]
> I'm guessing my original reasoning for not handling TRYAGAIN (if I knew about it at all) was that it's not clear how long a reshard will take, and so retrying again in a loop without any kind of wait in between wouldn't necessarily do anything except slow things down. It might be better to just keep returning it. And if it's being returned directly to the user when it happens, is it necessary to trace it?

I thought about adding a wait (similiar to how we could wait on `CLUSTERDOWN`) but after thinking about it a bit more, I agree that not handling `TRYAGAIN` seems like the better idea.

But hey, at least the stub cluster can now generate `TRYAGAIN` :-D

> Do you think it would be better to have Cluster delay all calls to Do as soon as a CLUSTERDOWN has been seen? Theoretically that would prevent the most disruption to the user, but it's also fairly drastic.

Yeah, I also thought about that. We could use a short default wait and allow users to change or disable the waiting. I'd guess that most users would prefer a short wait followed by a retry to just returning the error by default.

Basically on `CLUSTERDOWN` we would track the state as down, and retry once. In `doInner` we could than just check the state at the beginning and sleep for the configured time (if enabled). Since the 

> I'm not sure if we should trace on every occurrence of CLUSTERDOWN, since again if it's getting returned directly to the user the trace isn't giving them any information they don't already have. 

It allows for easier tracking, since it can be easier handled in one place. The errors returned by `Do` would need to be manually inspected for `CLUSTERDOWN` in some way.

We could also trace it only once if we track the cluster state and than maybe just trace whether the cluster is down or not.

**Comment by mediocregopher** [comment]
> But hey, at least the stub cluster can now generate TRYAGAIN :-D

lol, always a silver lining

> Basically on CLUSTERDOWN we would track the state as down, and retry once. In doInner we could than just check the state at the beginning and sleep for the configured time (if enabled).

That sounds good to me.

> We could also trace it only once if we track the cluster state and than maybe just trace whether the cluster is down or not.

Yeah, that's what I was thinking, I think that would be more useful than tracing every CLUSTERDOWN which gets seen.

**Comment by nussjustin** [comment]
Created a draft from my changes at #137.

For ease of prototyping I started with the simplest solutions for tracking the state and delaying the commands that came to my mind.
